From 601359a8e8ee3efb4fb6da5b06eed67c6b4217e3 Mon Sep 17 00:00:00 2001
From: Sachin Prabhu <sprabhu@redhat.com>
Date: Mon, 28 Jul 2014 12:16:37 -0400
Subject: [fs] cifs: Add protocol specific operation for CIFS xattrs

Message-id: <1406549822-10376-45-git-send-email-sprabhu@redhat.com>
Patchwork-id: 86778
O-Subject: [RHEL7.1 PATCH 44/69] [bz1082049]Add protocol specific operation for CIFS xattrs
Bugzilla: 1082049
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Steve French <smfrench@gmail.com>

Changeset 666753c3ef8fc88b0ddd5be4865d0aa66428ac35 added protocol
operations for get/setxattr to avoid calling cifs operations
on smb2/smb3 mounts for xattr operations and this changeset
adds the calls to cifs specific protocol operations for xattrs
(in order to reenable cifs support for xattrs which was
temporarily disabled by the previous changeset.  We do not
have SMB2/SMB3 worker function for setting xattrs yet so
this only enables it for cifs.

CCing stable since without these two small changsets (its
small coreq 666753c3ef8fc88b0ddd5be4865d0aa66428ac35 is
also needed) calling getfattr/setfattr on smb2/smb3 mounts
causes problems.

Signed-off-by: Steve French <smfrench@gmail.com>
Reviewed-by: Shirish Pargaonkar <spargaonkar@suse.com>
CC: Stable <stable@kernel.org>
(cherry picked from commit d979f3b0a1f0b5499ab85e68cdf02b56852918b6)
Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/cifs/inode.c b/fs/cifs/inode.c
index 62bf5a8..2aa8f53 100644
--- a/fs/cifs/inode.c
+++ b/fs/cifs/inode.c
@@ -527,10 +527,15 @@ static int cifs_sfu_mode(struct cifs_fattr *fattr, const unsigned char *path,
   return PTR_ERR(tlink);
  tcon = tlink_tcon(tlink);
 
- rc = CIFSSMBQAllEAs(xid, tcon, path, "SETFILEBITS",
-       ea_value, 4 /* size of buf */, cifs_sb->local_nls,
-       cifs_sb->mnt_cifs_flags &
-    CIFS_MOUNT_MAP_SPECIAL_CHR);
+ if (tcon->ses->server->ops->query_all_EAs == NULL) {
+  cifs_put_tlink(tlink);
+  return -EOPNOTSUPP;
+ }
+
+ rc = tcon->ses->server->ops->query_all_EAs(xid, tcon, path,
+   "SETFILEBITS", ea_value, 4 /* size of buf */,
+   cifs_sb->local_nls,
+   cifs_sb->mnt_cifs_flags & CIFS_MOUNT_MAP_SPECIAL_CHR);
  cifs_put_tlink(tlink);
  if (rc < 0)
   return (int)rc;
diff --git a/fs/cifs/smb1ops.c b/fs/cifs/smb1ops.c
index 40598b0..85fb460 100644
--- a/fs/cifs/smb1ops.c
+++ b/fs/cifs/smb1ops.c
@@ -1078,6 +1078,10 @@ struct smb_version_operations smb1_operations = {
  .query_mf_symlink = cifs_query_mf_symlink,
  .create_mf_symlink = cifs_create_mf_symlink,
  .is_read_op = cifs_is_read_op,
+#ifdef CONFIG_CIFS_XATTR
+ .query_all_EAs = CIFSSMBQAllEAs,
+ .set_EA = CIFSSMBSetEA,
+#endif /* CIFS_XATTR */
 };
 
 struct smb_version_values smb1_values = {
-- 
1.7.1